### PR TITLE
Validation for custom role

### PR DESCRIPTION
**Pull request by niravparikh05**
### What does this PR change?

- It adds validation for bare minimum role permissions that are required for custom roles.

### Does the PR depend on any other PRs or Issues? If yes, please list them.

- No

### Checklist

I confirm, that I have...

- [x] Read and followed the contributing guide in `CONTRIBUTING.md`
- [ ] Added tests for this PR
- [x] Formatted the code using `go fmt` (if applicable)
- [ ] Updated [documentation on the Paralus docs site](https://github.com/paralus/website/blob/main/docs) (if applicable)
- [x] Updated `CHANGELOG.md`

### CODE REVIEW

File: pkg/service/role.go [hunk 1]
```diff
@@ -145,6 +145,12 @@ func (s *roleService) Create(ctx context.Context, role *rolev3.Role) (*rolev3.Ro
 		}
 	}
 
+	//validate basic mandatory permissions that should be part of all custom roles
+	if len(role.Spec.Rolepermissions) > 0 && (!utils.Contains(role.Spec.Rolepermissions, partnerR) ||
+		!utils.Contains(role.Spec.Rolepermissions, organizationR) || !utils.Contains(role.Spec.Rolepermissions, projectR)) {
```
Comment (meain) [inline pkg/service/role.go:150]: Just a nit for better [readability](https://en.wikipedia.org/wiki/De_Morgan%27s_laws)
```suggestion
	if len(role.Spec.Rolepermissions) > 0 &&
		!(utils.Contains(role.Spec.Rolepermissions, partnerR) &&
			utils.Contains(role.Spec.Rolepermissions, organizationR) &&
			utils.Contains(role.Spec.Rolepermissions, projectR)) {
```

File: pkg/service/role.go [hunk 2]
```diff
@@ -264,6 +270,12 @@ func (s *roleService) Update(ctx context.Context, role *rolev3.Role) (*rolev3.Ro
 		}
 	}
 
+	//validate basic mandatory permissions that should be part of all custom roles
+	if !utils.Contains(role.Spec.Rolepermissions, partnerR) || !utils.Contains(role.Spec.Rolepermissions, organizationR) ||
+		!utils.Contains(role.Spec.Rolepermissions, projectR) {
```
Comment (meain) [inline pkg/service/role.go:275]: Small readability improvement
```suggestion
	if !(utils.Contains(role.Spec.Rolepermissions, partnerR) &&
		utils.Contains(role.Spec.Rolepermissions, organizationR) &&
		utils.Contains(role.Spec.Rolepermissions, projectR)) {
```

File: pkg/service/role_test.go [hunk 3]
```diff
@@ -155,15 +155,19 @@ func TestCreateRoleWithPermissions(t *testing.T) {
...
+		WithArgs().WillReturnRows(sqlmock.NewRows([]string{"id"}).AddRow(uuid.New().String()))
+	mock.ExpectQuery(`SELECT "resourcepermission"."id" FROM "authsrv_resourcepermission" AS "resourcepermission" WHERE .name = 'project.read'.`).
 		WithArgs().WillReturnRows(sqlmock.NewRows([]string{"id"}).AddRow(uuid.New().String()))
 	mock.ExpectQuery(`INSERT INTO "authsrv_resourcerolepermission"`).
 		WithArgs().WillReturnRows(sqlmock.NewRows([]string{"id"}).AddRow(uuid.New().String()))
 	mock.ExpectCommit()
 
 	role := &rolev3.Role{
 		Metadata: &v3.Metadata{Partner: "partner-" + puuid, Organization: "org-" + ouuid, Name: "role-" + ruuid},
-		Spec:     &rolev3.RoleSpec{IsGlobal: true, Scope: "system", Rolepermissions: []string{"ops_star.all"}},
```
Comment (meain) [inline pkg/service/role_test.go:166]: Should we have separate check for `ops_star.all` as it is a catchall?

**Review by meain — COMMENTED** [review]
Although it does not makes sense to create user which needs to use the UI without these permissions I feel like we should probably keep it open so that they can create a role without this for api only access. What do you think?

*One other things we could do is to hint in the UI that they will need to choose these roles if they `are` creating a user who is going to use Paralus dashboard.*

**Comment by niravparikh05** [comment]
> Although it does not makes sense to create user which needs to use the UI without these permissions I feel like we should probably keep it open so that they can create a role without this for api only access. What do you think?
> 
> _One other things we could do is to hint in the UI that they will need to choose these roles if they `are` creating a user who is going to use Paralus dashboard._

also, as discussed removed project.read being mandatory to support custom role keeping api only access usage in mind. However retained org.read and partner.read as these are mandatory for all dashboard / api use-cases and until we make these api endpoints open for all by default.

**Comment by niravparikh05** [comment]
@meain As discussed, can we go ahead with current restriction on Partner / Organization permissions for both dashboard and api ( system users ) as these would be required in order to invoke most other dependent apis until we plan to make both these endpoints public and non-restricted.

File: pkg/service/const.go [hunk 4]
```diff
@@ -10,4 +10,7 @@ const (
 	projectScope   = "project"
 	namespaceR     = "kubectl.namespace.read"
 	namespaceW     = "kubectl.namespace.write"
+	partnerR       = "partner.read"
+	organizationR  = "organization.read"
+	projectR       = "project.read"
```
Comment (meain) [inline pkg/service/const.go:15]: We can probably drop this as this won't be used anywhere.

File: pkg/service/role.go [hunk 5]
```diff
@@ -264,6 +271,12 @@ func (s *roleService) Update(ctx context.Context, role *rolev3.Role) (*rolev3.Ro
 		}
 	}
 
+	//validate basic mandatory permissions that should be part of all custom roles
+	if !(utils.Contains(role.Spec.Rolepermissions, partnerR) &&
+		utils.Contains(role.Spec.Rolepermissions, organizationR)) {
+		return nil, fmt.Errorf("invalid role permissions, '%v', '%v', '%v' should be present ", partnerR, organizationR, projectR)
```
Comment (meain) [inline pkg/service/role.go:277]: We should probably drop the project check from here as we don't necessarily need it as of now.

**Review by meain — COMMENTED** [review]
Few minor tweaks

File: pkg/service/role_test.go (hunk 3 above)
Comment (meain) [reply pkg/service/role_test.go:166]: As discussed, this is not being used and will be removed soon.

**Review by meain — APPROVED** [review]
LGTM

**Pull request merged**